File: android/src/main/java/com/dynamsoft/capacitor/dce/CameraPreviewPlugin.java
package com.dynamsoft.capacitor.dce;

import android.Manifest;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.util.Base64;
import android.util.Log;
import android.util.Size;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.dynamsoft.core.CoreException;
import com.dynamsoft.dce.CameraEnhancer;
import com.dynamsoft.dce.CameraEnhancerException;
import com.dynamsoft.dce.DCECameraView;
import com.dynamsoft.dce.DCEPhotoListener;
import com.dynamsoft.dce.EnumCameraState;
import com.dynamsoft.dce.EnumResolution;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

import java.io.ByteArrayOutputStream;

@CapacitorPlugin(
    name = "CameraPreview",
    permissions = {
        @Permission(strings = { Manifest.permission.CAMERA }, alias = CameraPreviewPlugin.CAMERA),
    }
)
public class CameraPreviewPlugin extends Plugin {
    // Permission alias constants
    static final String CAMERA = "camera";
    private CameraEnhancer mCameraEnhancer;
    private DCECameraView mCameraView;
    private EnumCameraState previousCameraStatus;
    private String callbackID;

    @PluginMethod
    public void initialize(PluginCall call) {
        getActivity().runOnUiThread(new Runnable() {
            public void run() {
                mCameraEnhancer = new CameraEnhancer(getActivity());
                mCameraView = new DCECameraView(getActivity());
                mCameraEnhancer.setCameraView(mCameraView);
                FrameLayout.LayoutParams cameraPreviewParams = new FrameLayout.LayoutParams(
                        FrameLayout.LayoutParams.WRAP_CONTENT,
                        FrameLayout.LayoutParams.WRAP_CONTENT
                );
                ((ViewGroup) bridge.getWebView().getParent()).addView(mCameraView,cameraPreviewParams);
                bridge.getWebView().bringToFront();
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void startCamera(PluginCall call) {
        getActivity().runOnUiThread(new Runnable() {
            public void run() {
                try {
                    mCameraView.setVisibility(View.VISIBLE);
                    mCameraEnhancer.open();
                    makeWebViewTransparent();
                    triggerOnPlayed();
                    call.resolve();
                } catch (CameraEnhancerException e) {
                    e.printStackTrace();
                    call.reject(e.getMessage());
                }
            }
        });
    }

    @PluginMethod
    public void stopCamera(PluginCall call) {
        try{
            restoreWebViewBackground();
            mCameraView.setVisibility(View.INVISIBLE);
            mCameraEnhancer.close();
            call.resolve();
        }catch (Exception e){
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void pauseScan(PluginCall call) {
        try{
            mCameraEnhancer.pause();
            call.resolve();
        }catch (Exception e){
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void resumeScan(PluginCall call) {
        try{
            mCameraEnhancer.resume();
            call.resolve();
        }catch (Exception e){
            call.reject(e.getMessage());
        }
    }


    private void makeWebViewTransparent(){
        bridge.getWebView().setTag(bridge.getWebView().getBackground());
        bridge.getWebView().setBackgroundColor(Color.TRANSPARENT);
    }

    private void restoreWebViewBackground(){
        bridge.getWebView().setBackground((Drawable) bridge.getWebView().getTag());
    }

    @PluginMethod
    public void toggleTorch(PluginCall call) {
        try{
            if (call.getBoolean("on",true)){
                mCameraEnhancer.turnOnTorch();
            }else {
                mCameraEnhancer.turnOffTorch();
            }
            call.resolve();
        }catch (Exception e){
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void setScanRegion(PluginCall call){
        if (mCameraEnhancer!=null) {
            getActivity().runOnUiThread(new Runnable() {
                public void run() {
                    try {
                        JSObject region = call.getObject("region");
                        com.dynamsoft.core.RegionDefinition scanRegion = new com.dynamsoft.core.RegionDefinition();
                        scanRegion.regionTop = region.getInt("top");
                        scanRegion.regionBottom = region.getInt("bottom");
                        scanRegion.regionLeft = region.getInt("left");
                        scanRegion.regionRight = region.getInt("right");
                        scanRegion.regionMeasuredByPercentage = region.getInt("measuredByPercentage");

                        mCameraEnhancer.setScanRegion(scanRegion);
                        call.resolve();
                    } catch (Exception e) {
                        e.printStackTrace();
                        call.reject(e.getMessage());
                    }
                }
            });
        }else{
            call.reject("DCE not initialized");
        }
    }

    @PluginMethod
    public void setZoom(PluginCall call){
        if (call.hasOption("factor")) {
            Float factor = call.getFloat("factor");
            try {
                mCameraEnhancer.setZoom(factor);
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }
        call.resolve();
    }

    @PluginMethod
    public void setFocus(PluginCall call){
        if (call.hasOption("x") && call.hasOption("y")) {
            Float x = call.getFloat("x");
            Float y = call.getFloat("y");
            try {
                mCameraEnhancer.setFocus(x,y);
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }
        call.resolve();
    }

    @PluginMethod
    public void selectCamera(PluginCall call){
        if (call.hasOption("cameraID")){
            try {
                Runnable selectCameraRunnable = new Runnable() {
                    public void run() {
                        try {
                            mCameraEnhancer.selectCamera(call.getString("cameraID"));
                        } catch (CameraEnhancerException e) {
                            e.printStackTrace();
                        }
                        synchronized(this)
                        {
                            this.notify();
                        }
                    }
                };
                synchronized( selectCameraRunnable ) {
                    getActivity().runOnUiThread(selectCameraRunnable);
                    selectCameraRunnable.wait();
                }

                triggerOnPlayed();
            } catch (InterruptedException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }
        JSObject result = new JSObject();
        result.put("success",true);
        call.resolve(result);
    }

    private void triggerOnPlayed(){
        Size res = mCameraEnhancer.getResolution();
        if (res != null) {
            JSObject onPlayedResult = new JSObject();
            onPlayedResult.put("resolution",res.getWidth() + "x" + res.getHeight());
            Log.d("DBR","resolution:" + res.getWidth() + "x" + res.getHeight());
            notifyListeners("onPlayed",onPlayedResult);
        }
    }

    @PluginMethod
    public void getAllCameras(PluginCall call){
        if (mCameraEnhancer == null) {
            call.reject("not initialized");
        }else {
            JSObject result = new JSObject();
            JSArray cameras = new JSArray();
            for (String camera: mCameraEnhancer.getAllCameras()) {
                cameras.put(camera);
            }
            result.put("cameras",cameras);
            call.resolve(result);
        }
    }

    @PluginMethod
    public void getSelectedCamera(PluginCall call){
        if (mCameraEnhancer == null) {
            call.reject("not initialized");
        }else{
            JSObject result = new JSObject();
            result.put("selectedCamera",mCameraEnhancer.getSelectedCamera());
            call.resolve(result);
        }
    }

    @PluginMethod
    public void setResolution(PluginCall call){
        if (call.hasOption("resolution")){
            try {
                Runnable setResolutionRunnable = new Runnable() {
                    public void run() {
                        try {
                            mCameraEnhancer.setResolution(EnumResolution.fromValue(call.getInt("resolution")));
                        } catch (CameraEnhancerException e) {
                            e.printStackTrace();
                        }
                        synchronized(this)
                        {
                            this.notify();
                        }
                    }
                };
                synchronized( setResolutionRunnable ) {
                    getActivity().runOnUiThread(setResolutionRunnable);
                    setResolutionRunnable.wait();
                }
                triggerOnPlayed();
            } catch (InterruptedException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }
        JSObject result = new JSObject();
        result.put("success",true);
        call.resolve(result);
    }

    @PluginMethod
    public void getResolution(PluginCall call){
        if (mCameraEnhancer == null) {
            call.reject("DCE not initialized");
        }else{
            String res = mCameraEnhancer.getResolution().getWidth()+"x"+mCameraEnhancer.getResolution().getHeight();
            JSObject result = new JSObject();
            result.put("resolution",res);
            call.resolve(result);
        }
    }

    @PluginMethod
    public void takeSnapshot(PluginCall call){
        try {
            if (mCameraEnhancer.getCameraState() == EnumCameraState.OPENED) {
                Bitmap bitmap = mCameraEnhancer.getFrameFromBuffer(true).toBitmap();
                String base64 = bitmap2Base64(bitmap);
                JSObject result = new JSObject();
                result.put("base64",base64);
                call.resolve(result);
            }else{
                call.reject("camera is not open");
            }
        } catch (Exception e) {
            e.printStackTrace();
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void takePhoto(PluginCall call){
        mCameraEnhancer.takePhoto(new DCEPhotoListener() {
            @Override
            public void photoOutputCallback(byte[] bytes) {
                String base64 = Base64.encodeToString(bytes, Base64.DEFAULT);
                JSObject result = new JSObject();
                result.put("base64",base64);
                call.resolve(result);
            }
        });
    }

    public static String bitmap2Base64(Bitmap bitmap) {
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream);
        return Base64.encodeToString(outputStream.toByteArray(), Base64.DEFAULT);
    }

    @PluginMethod
    public void isOpen(PluginCall call){
        if (mCameraEnhancer != null) {
            JSObject result = new JSObject();
            if (mCameraEnhancer.getCameraState() == EnumCameraState.OPENED) {
                result.put("isOpen",true);
            }else{
                result.put("isOpen",false);
            }
            call.resolve(result);
        }else {
            call.reject("DCE not initialized.");
        }
    }

    @Override
    protected void handleOnPause() {
        if (mCameraEnhancer!=null){
            try {
                previousCameraStatus = mCameraEnhancer.getCameraState();
                mCameraEnhancer.close();
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
            }
        }
        super.handleOnPause();
    }

    @Override
    protected void handleOnResume() {
        if (mCameraEnhancer!=null){
            try {
                if (previousCameraStatus == EnumCameraState.OPENED) {
                    mCameraEnhancer.open();
                }
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
            }
        }
        super.handleOnResume();
    }

    @PluginMethod
    public void requestCameraPermission(PluginCall call) {
        boolean hasCameraPerms = getPermissionState(CAMERA) == PermissionState.GRANTED;
        if (hasCameraPerms == false) {
            Log.d("DBR","no camera permission. request permission.");
            String[] aliases = new String[] { CAMERA };
            requestPermissionForAliases(aliases, call, "cameraPermissionsCallback");
        }else{
            call.resolve();
        }
    }

    @PermissionCallback
    private void cameraPermissionsCallback(PluginCall call) {
        boolean hasCameraPerms = getPermissionState(CAMERA) == PermissionState.GRANTED;
        if (hasCameraPerms) {
            call.resolve();
        }else {
            call.reject("Permission not granted.");
        }
    }
}

File: ios/Plugin/CameraPreviewPlugin.swift
import Foundation
import Capacitor
import DynamsoftCameraEnhancer
/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CameraPreviewPlugin)
public class CameraPreviewPlugin: CAPPlugin {
    static var dce:DynamsoftCameraEnhancer! = nil
    var dceView:DCECameraView! = nil
    @objc func initialize(_ call: CAPPluginCall) {
        // Initialize a camera view for previewing video.
        DispatchQueue.main.sync {
            dceView = DCECameraView.init(frame: (bridge?.viewController?.view.bounds)!)
            self.webView!.superview!.insertSubview(dceView, belowSubview: self.webView!)
            CameraPreviewPlugin.dce = DynamsoftCameraEnhancer.init(view: dceView)
            CameraPreviewPlugin.dce.setResolution(EnumResolution.EnumRESOLUTION_720P)
        }
        call.resolve()
    }
    
    @objc func startCamera(_ call: CAPPluginCall) {
        makeWebViewTransparent()
        if CameraPreviewPlugin.dce != nil {
            DispatchQueue.main.sync {
                CameraPreviewPlugin.dce.open()
                triggerOnPlayed()
            }
        }else{
            call.reject("DCE not initialized")
            return
        }
        call.resolve()
    }
    
    func makeWebViewTransparent(){
        DispatchQueue.main.async {
           self.bridge?.webView!.isOpaque = false
           self.bridge?.webView!.backgroundColor = UIColor.clear
           self.bridge?.webView!.scrollView.backgroundColor = UIColor.clear
       }
    }
    func restoreWebViewBackground(){
        DispatchQueue.main.async {
           self.bridge?.webView!.isOpaque = true
           self.bridge?.webView!.backgroundColor = UIColor.white
           self.bridge?.webView!.scrollView.backgroundColor = UIColor.white
       }
    }
    
    @objc func toggleTorch(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            DispatchQueue.main.sync {
                if call.getBool("on", true){
                    CameraPreviewPlugin.dce.turnOnTorch()
                } else{
                    CameraPreviewPlugin.dce.turnOffTorch()
                }
            }
            call.resolve()
        }
    }
    
    @objc func stopCamera(_ call: CAPPluginCall) {
        restoreWebViewBackground()
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            DispatchQueue.main.sync {
                CameraPreviewPlugin.dce.close()
            }
            call.resolve()
        }
    }
    
    @objc func resumeCamera(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            CameraPreviewPlugin.dce.resume()
            call.resolve()
        }
    }
    
    @objc func pauseCamera(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            CameraPreviewPlugin.dce.pause()
            call.resolve()
        }
    }
    
    @objc func setResolution(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            let res = call.getInt("resolution") ?? -1
            NSLog("Resolution: %d", res)
            
            if res != -1 {
                let resolution = EnumResolution.init(rawValue: res)
                CameraPreviewPlugin.dce.setResolution(resolution!)
                triggerOnPlayed()
            }
            call.resolve()
        }
    }
    
    @objc func getResolution(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            var ret = PluginCallResultData()
            let res = CameraPreviewPlugin.dce.getResolution();
            CameraPreviewPlugin.dce.getResolution()
            print("res: "+res)
            ret["resolution"] = res
            call.resolve(ret)
        }
    }
    
    @objc func triggerOnPlayed() {
        if (CameraPreviewPlugin.dce != nil) {
            var ret = PluginCallResultData()
            let res = CameraPreviewPlugin.dce.getResolution()
            ret["resolution"] = res
            print("trigger on played")
            notifyListeners("onPlayed", data: ret)
        }
    }
    
    @objc func getAllCameras(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            var ret = PluginCallResultData()
            let array = NSMutableArray();
            array.addObjects(from: CameraPreviewPlugin.dce.getAllCameras())
            ret["cameras"] = array
            call.resolve(ret)
        }
    }
    
    @objc func getSelectedCamera(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            var ret = PluginCallResultData()
            ret["selectedCamera"] = CameraPreviewPlugin.dce.getSelectedCamera()
            call.resolve(ret)
        }
    }
    
    @objc func selectCamera(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            let cameraID = call.getString("cameraID") ?? ""
            if cameraID != "" {
                try? CameraPreviewPlugin.dce.selectCamera(cameraID)
                triggerOnPlayed()
            }
            call.resolve()
        }
    }
    
    @objc func setScanRegion(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            let region = call.getObject("region")
            let scanRegion = iRegionDefinition()
            scanRegion.regionTop = region?["top"] as! Int
            scanRegion.regionBottom = region?["bottom"] as! Int
            scanRegion.regionLeft = region?["left"] as! Int
            scanRegion.regionRight = region?["right"] as! Int
            scanRegion.regionMeasuredByPercentage = region?["measuredByPercentage"] as! Int
            try? CameraPreviewPlugin.dce.setScanRegion(scanRegion)
            call.resolve()
        }
    }
    
    @objc func setZoom(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            let factor:CGFloat = CGFloat(call.getFloat("factor") ?? 1.0)
            CameraPreviewPlugin.dce.setZoom(factor)
            call.resolve()
        }
    }
    
    @objc func setFocus(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            let x = call.getFloat("x", -1.0);
            let y = call.getFloat("y", -1.0);
            if x != -1.0 && y != -1.0 {
                CameraPreviewPlugin.dce.setFocus(CGPoint(x: CGFloat(x), y: CGFloat(y)))
            }
            call.resolve()
        }
    }
    
    @objc func requestCameraPermission(_ call: CAPPluginCall) {
        call.resolve()
    }
    
    @objc func isOpen(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            var ret = PluginCallResultData()
            if CameraPreviewPlugin.dce.getCameraState() == EnumCameraState.EnumCAMERA_STATE_OPENED {
                ret["isOpen"] = true
            } else {
                ret["isOpen"] = false
            }
            call.resolve(ret)
        }
    }
    
    @objc static func getBitmap() -> UIImage? {
        if (CameraPreviewPlugin.dce == nil){
            return nil
        }else{
            let frame = CameraPreviewPlugin.dce.getFrameFromBuffer(true)
            var ret = PluginCallResultData()
            if let img = frame.toUIImage() {
                var cropped:UIImage
                if frame.isCropped {
                    cropped = croppedUIImage(image: img, region: CameraPreviewPlugin.dce.getScanRegion(),degree: frame.orientation)
                } else {
                    cropped = img
                }
                let rotated = rotatedUIImage(image: cropped, degree: frame.orientation)
                let normalized = normalizedImage(rotated);
                return normalized
            }else{
                return nil
            }
        }
    }
    
    @objc func takeSnapshot(_ call: CAPPluginCall) {
        if (CameraPreviewPlugin.dce == nil){
            call.reject("DCE not initialized")
        }else{
            let quality = call.getInt("quality",85)
            let frame = CameraPreviewPlugin.dce.getFrameFromBuffer(true)
            
            var ret = PluginCallResultData()
            if let img = frame.toUIImage() {
                var cropped:UIImage
                if frame.isCropped {
                    cropped = CameraPreviewPlugin.croppedUIImage(image: img, region: CameraPreviewPlugin.dce.getScanRegion(),degree: frame.orientation)
                } else {
                    cropped = img
                }
                let rotated = CameraPreviewPlugin.rotatedUIImage(image: cropped, degree: frame.orientation)
                let normalized = CameraPreviewPlugin.normalizedImage(rotated);
                let base64 = getBase64FromImage(image: normalized, quality: CGFloat(quality/100))
                ret["base64"] = base64
                call.resolve(ret)
            }else{
                call.reject("Failed to take a snapshot")
            }
        }
    }
    
    static func rotatedUIImage(image:UIImage, degree: Int) -> UIImage {
        var rotatedImage = UIImage()
        switch degree
        {
            case 90:
                rotatedImage = UIImage(cgImage: image.cgImage!, scale: 1.0, orientation: .right)
            case 180:
                rotatedImage = UIImage(cgImage: image.cgImage!, scale: 1.0, orientation: .down)
            default:
                return image
        }
        return rotatedImage
    }
    
    static func croppedUIImage(image:UIImage, region:iRegionDefinition, degree: Int) -> UIImage {
        let cgImage = image.cgImage
        let imgWidth = Double(cgImage!.width)
        let imgHeight = Double(cgImage!.height)
        
        var regionLeft = Double(region.regionLeft) / 100.0
        var regionTop = Double(region.regionTop) / 100.0
        var regionWidth = Double(region.regionRight - region.regionLeft) / 100.0
        var regionHeight = Double(region.regionBottom - region.regionTop) / 100.0

        if degree == 90 {
            let temp1 = regionLeft
            regionLeft = regionTop
            regionTop = temp1
            let temp2 = regionWidth
            regionWidth = regionHeight
            regionHeight = temp2
            print("degree 90")
        }else if degree == 180 {
            regionTop = 1.0 - regionTop
            print("degree 180")
        }
        let left:Double = regionLeft * imgWidth
        let top:Double = regionTop * imgHeight
        let width:Double = regionWidth * imgWidth
        let height:Double = regionHeight * imgHeight
        //print("imgWidth")
        //print(imgWidth)
        //print("regionTop")
        //print(regionTop)
        //print("top")
        //print(top)
        //print("regionleft")
        //print(regionLeft)
        //print("left")
        //print(left)
        //print("regionWidth")
        //print(regionWidth)
        //print("regionHeight")
        //print(regionHeight)
        //print("width")
        //print(width)
        //print("height")
        //print(height)
        
        // The cropRect is the rect of the image to keep,
        // in this case centered
        let cropRect = CGRect(
            x: left,
            y: top,
            width: width,
            height: height
        ).integral

        let cropped = cgImage?.cropping(
            to: cropRect
        )!
        let image = UIImage(cgImage: cropped!)
        return image
    }
    
    static func normalizedImage(_ image:UIImage) -> UIImage {
        if image.imageOrientation == UIImage.Orientation.up {
            return image
        }
        UIGraphicsBeginImageContextWithOptions(image.size, false, image.scale)
        image.draw(in: CGRect(x:0,y:0,width:image.size.width,height:image.size.height))
        let normalized = UIGraphicsGetImageFromCurrentImageContext()!
        UIGraphicsEndImageContext();
        return normalized
    }
    
    func getBase64FromImage(image:UIImage, quality: CGFloat) -> String{
       let dataTmp = image.jpegData(compressionQuality: quality)
       if let data = dataTmp {
           return data.base64EncodedString()
       }
       return ""
    }
    
    @objc func takePhoto(_ call: CAPPluginCall) {
        takeSnapshot(call)
    }
    
}
